pyprep/utils.py

```diff
@@ -325,6 +321,31 @@ def _correlate_arrays(a, b, matlab_strict=False):
         return np.diag(np.corrcoef(a, b)[:n_chan, n_chan:])
 
 
+def madmed(x, axis=None):
+    """Calculate median absolute deviations from the median (MAD) for an array.
+
+    Parameters
+    ----------
+    x : np.ndarray
+        Input array containing the values to summarize.
+    axis : {int, tuple of int, None}, optional
+        Axis along which MADs should be calculated. If ``None``, the MAD will
+        be calculated for the full input array. Defaults to ``None``.
+
+    Returns
+    -------
+    mad : scalar or np.ndarray
+        If no axis is specified, returns the MAD for the full input array as a
+        single numeric value. Otherwise, returns an ``np.ndarray`` containing
+        the MAD for each index along the specified axis.
+
+    """
+    med = np.median(x, axis=axis)
+    if axis == 1:
+        med = med.reshape(-1, 1)  # Transposes array to allow subtraction below
+    return np.median(np.abs(x - med), axis=axis)
```

Comment:
Agreed with 2) and 3) - should all the functions in utils.py be renamed with the `_` prefix, then? Most are already, but there are several at the bottom that aren't. Since `utils.py` is already ignored by sphinxdoc I wasn't sure if it mattered, but the `_` prefix certainly makes it more explicit.

As to 4), the normalization constant does the same thing that `MAD_TO_SD` does: scales the MAD to be on the same unit scale as the standard deviation (assuming population normality). In the old code, we always set the normalization constant to 1 (i.e., no scaling) and then scaled using the `MAD_TO_SD` number where appropriate. Will save a cycle or two avoiding an unnecessary multiplication by 1!

5) I haven't tested it against MATLAB's `mad` explicitly, but given that it produces identical values in the MATLAB comparison tests I'm assuming that it's equivalent for our purposes.

6) Haven't tested a >2D array (not sure if it'd work), so I'll limit it to 2D arrays and add a check for that in the code.

7) Ahh yep, will fix in my revisions.

Reply:
>  should all the functions in utils.py be renamed with the _ prefix, then?

yes, I think so. Apart from making it explicit, it also influences the import inspection behavior I think (when users use auto-completion, for example)

> As to 4), the normalization constant does the same thing that MAD_TO_SD does: scales the MAD to be on the same unit scale as the standard deviation

alright, thanks!

>  but given that it produces identical values in the MATLAB comparison tests I'm assuming that it's equivalent for our purposes.

agreed

6 & 7: Okay good, thanks for fixing it.

Reply:
Okay, all of this should be fixed in the latest commit! For the utils.py `_` thing, I also ended up moving one function from `utils.py` to `tests/conftest.py` since it's only ever used/intended for use in the test suite.